### ISSUE TITLE
Stop-based ray aiming

### ISSUE THREAD

**Issue by alfredclwong** [microsoft/OpticSim.jl#178]
Andreas frequently mentions that it would be nice to filter rays based on angle and location of incidence with a stop surface. Apparently, this is often used in a backtraced manner to construct an emitter which satisfies the constraints.

Have we had plans for this in the past?

**Comment by friggog** [comment]
We discussed this in the past, I think the conclusion we came to at the time was that it was cheap enough just to put the stop there and fire rays in (almost) any direction. The bad rays will get blocked by the stop almost immediately and not cost a lot. 

It would be pretty easy to add some basic ray aiming, particularly in the axisymmetric case - when you got to more complex systems it rapidly gets very complicated, more in how you specify what surface you want to aim at, rather than the actual aiming of the rays.

TLDR; shooting rays is probably cheap enough that we shouldn't worry about this from a performance standpoint, if there is other motivation to do it then I'm all for it

**Comment by BrianGun** [comment]
I assume this is for computing a stop or baffle that will block rays are going where they shouldn't. With the trackallrays option you can get the trace function to return all the rays along one path for each original ray that is shot into the system. 

Then you can examine the rays along that one path to filter them any way you want. For example, if you wanted to allow only rays that went from the emitter and intersected just a certain set of lens elements or surfaces along the way then you would search through the path data structure and if you found one of the forbidden surfaces you'd return a bool indicating that the ray should be filtered out.

Doesn't seem like a _feature_ we need to add since it should be easy for the user to code from scratch with what we already have and there are likely to be many different types of filtering people want to do. 

Would make a good example to add to the documentation though so people understand how to use a powerful feature that is now obscure. And it might be that once we start doing this kind of ray filtering a lot we will discover some utility functions would be useful for common cases.

**Comment by alfredclwong** [comment]
Let's leave this for now then.
